### PR TITLE
Add option to write a log-file

### PR DESCRIPTION
**Pull request by mcwimm**
This is an option - although not the most elegant - to write a log file.

This can be done using `-l [project-file]` instead of `-i [project-file]` to run the model. 
If activated a file named `MANGA.log` is saved to the root directory.

The log file contains: time, class name, method name and `self`, although in a reverse order if a method is inherited (I don't know why).

Logs are only generated for methods with the decorator `@method_logger`. The decorator must be added to all upstream methods (parents, grandparents) to log all calls.
Currently, the decorator is added to all methods in below-ground and growth-and-death concepts containing inheritance.

The script `Logger.py` contains also other options that can be used to adapt or specify log-file generation.
@jbathmann : currently, `Logger.py` is located in `TreeModelLib` , which is not the best location yet. Do you have an idea where it makes the most sense? Root directory?

Note: It is reasonable to calculate only one or a few time steps when logging is enabled.

### CODE REVIEW

File: TreeModelLib/Logger.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+'''
+The function and class below provide a method to logg all calls (but only 
+the first level of their occurrence, i.e. without inheritance).
+
+Change required in TreeModel to: 
+
+    from .Logger import MethodLogger
+    class TreeModel(MethodLogger):
+'''
```
Comment (jbathmann) [inline TreeModelLib/Logger.py:35]: I do not get this. When do you use it?

File: TreeModelLib/BelowgroundCompetition/NetworkFixedSalinity/NetworkFixedSalinity.py [hunk 2]
```diff
@@ -8,6 +8,7 @@
 import numpy as np
 from TreeModelLib.BelowgroundCompetition.SimpleNetwork import SimpleNetwork
 from TreeModelLib.BelowgroundCompetition.FixedSalinity import FixedSalinity
+from TreeModelLib.Logger import method_logger
```
Comment (jbathmann) [inline TreeModelLib/BelowgroundCompetition/NetworkFixedSalinity/NetworkFixedSalinity.py:11]: Please update for other concepts.
Is it possible to add @method_logger to all concepts/methods automatically?
Which methods do you usually "decorate"?

File: MANGA.py [hunk 3]
```diff
@@ -52,6 +51,13 @@ def main(argv):
             sys.exit()
         elif opt in ("-i", "--project_file"):
             project_file = str(arg)
+        elif opt in ("-l", "--project_file"):
```
Comment (jbathmann) [inline MANGA.py:54]: I do not understand this. What happens, if MANGA is executed using
`python MANGA.py --project_file blabliblubb`

Shouldn't this line be 
```suggestion
        elif opt in ("-l", "--logging"):
```

File: MANGA.py [hunk 4]
```diff
@@ -52,6 +51,13 @@ def main(argv):
...
+            project_file = str(arg)
```
Comment (jbathmann) [inline MANGA.py:55]: delete?

File: TreeModelLib/Logger.py (hunk 1 above)
Comment (mcwimm) [reply TreeModelLib/Logger.py:35]: In case you want to log ALL calls you can modify `TreeModel` as in the example.

File: MANGA.py (hunk 4 above)
Comment (mcwimm) [reply MANGA.py:55]: see above

Comment (jbathmann) [reply MANGA.py:55]: delete? ;)

**Comment by jbathmann** [comment]
> The script Logger.py contains also other options that can be used to adapt or specify log-file generation.
> @jbathmann : currently, Logger.py is located in TreeModelLib , which is not the best location yet. Do you have an idea where it 
> makes the most sense? Root directory?

Can you add it to the processLib as it is defining how the process works in general?
Do you think this is reasonable?

**Comment by mcwimm** [comment]
> > The script Logger.py contains also other options that can be used to adapt or specify log-file generation.
> > @jbathmann : currently, Logger.py is located in TreeModelLib , which is not the best location yet. Do you have an idea where it
> > makes the most sense? Root directory?
> 
> Can you add it to the processLib as it is defining how the process works in general? Do you think this is reasonable?

So we create a new Lib?

**Comment by jbathmann** [comment]
> > > The script Logger.py contains also other options that can be used to adapt or specify log-file generation.
> > > @jbathmann : currently, Logger.py is located in TreeModelLib , which is not the best location yet. Do you have an idea where it
> > > makes the most sense? Root directory?
> > 
> > 
> > Can you add it to the processLib as it is defining how the process works in general? Do you think this is reasonable?
> 
> So we create a new Lib?

There already is a processLib

File: MANGA.py (hunk 3 above)
Comment (jbathmann) [reply MANGA.py:54]: my_project_file = dummy/for/path/to/my/project/file

Comment (mcwimm) [reply MANGA.py]: solved

File: MANGA.py (hunk 4 above)
Comment (mcwimm) [reply MANGA.py]: solved

**Comment by mcwimm** [comment]
> > > > The script Logger.py contains also other options that can be used to adapt or specify log-file generation.
> > > > @jbathmann : currently, Logger.py is located in TreeModelLib , which is not the best location yet. Do you have an idea where it
> > > > makes the most sense? Root directory?
> > > 
> > > 
> > > Can you add it to the processLib as it is defining how the process works in general? Do you think this is reasonable?
> > 
> > 
> > So we create a new Lib?
> 
> There already is a processLib

Can't find such Lib in the master. Am I blind?

**Comment by jbathmann** [comment]
> > > > > The script Logger.py contains also other options that can be used to adapt or specify log-file generation.
> > > > > @jbathmann : currently, Logger.py is located in TreeModelLib , which is not the best location yet. Do you have an idea where it
> > > > > makes the most sense? Root directory?
> > > > 
> > > > 
> > > > Can you add it to the processLib as it is defining how the process works in general? Do you think this is reasonable?
> > > 
> > > 
> > > So we create a new Lib?
> > 
> > 
> > There already is a processLib
> 
> Can't find such Lib in the master. Am I blind?

Of course not.  I was talking about the ProjectLib ;)

**Comment by mcwimm** [comment]
> > > > > > The script Logger.py contains also other options that can be used to adapt or specify log-file generation.
> > > > > > @jbathmann : currently, Logger.py is located in TreeModelLib , which is not the best location yet. Do you have an idea where it
> > > > > > makes the most sense? Root directory?
> > > > > 
> > > > > 
> > > > > Can you add it to the processLib as it is defining how the process works in general? Do you think this is reasonable?
> > > > 
> > > > 
> > > > So we create a new Lib?
> > > 
> > > 
> > > There already is a processLib
> > 
> > 
> > Can't find such Lib in the master. Am I blind?
> 
> Of course not. I was talking about the ProjectLib ;)

Ah, ok. File is moved and all imports are updated.

**Pull request merged**